### PR TITLE
[Perf]MOSS-TTS-Local-v1.5  Stage Transfer and Chunking

### PR DESCRIPTION
**Pull request by nagisa-kunhah**
## Purpose

#4676

This PR keeps the effective MOSS-TTS Local changes only:

- Stage-0 sends codec chunks in a layout that Stage-1 can consume reliably.
- Omni connector tensor serialization now handles tensors that are contiguous
  but whose trailing dimension has a stride incompatible with `view(torch.uint8)`.

## Test Plan

**vLLM Version:**

- vLLM: `0.24.0`
- PyTorch: `2.11.0+cu130`
- CUDA: `13.0`
- GPU: `NVIDIA H20`
- Python: `/root/vllm/.venv/bin/python`

**vLLM-Omni Commit:**

- `e60fd5d1cfd79a2f102131f8ebcd468cdaa73042`

**Runtime Configuration:**

- Model: `OpenMOSS-Team/MOSS-TTS-Local-Transformer-v1.5`
- Dataset: `/tmp/moss_local_history_long`
- Request path: `/v1/audio/speech`
- `max_new_tokens`: `1000`
- Server flags:
  - `--attention-backend TRITON_ATTN`
  - `--enforce-eager`
  - `--compilation-config '{"custom_ops":["all","-silu_and_mul"]}'`
  - `--ir-op-priority '{"rms_norm":["native"],"fused_add_rms_norm":["native"]}'`

## Test Result

Unit tests:

```bash
PYTHONPATH=/root/autodl-tmp/vllm_omni_bench/compat:/root/autodl-tmp/vllm_omni_bench/python_pkgs:/root/vllm-omni:/root/vllm \
/root/vllm/.venv/bin/python -m pytest -q \
  tests/distributed/omni_connectors/test_basic_connectors.py::test_noncontiguous_tensor_serialization \
  tests/distributed/omni_connectors/test_basic_connectors.py::test_size_one_trailing_dim_tensor_serialization \
  tests/model_executor/stage_input_processors/test_moss_tts_async_chunk.py
```

Result:

```text
14 passed, 17 warnings in 0.45s
```

Long-form E2E speech test:

```text
runs: 3
completed: 3
failed: 0
mean_audio_duration_s: 80.0
mean_e2el_ms: 49772.06
median_e2el_ms: 45565.51
mean_audio_rtf: 0.6222
median_audio_rtf: 0.5696
audio_throughput: 1.6311
```

No empty-audio responses, no server crash, and no
`SharedMemoryConnector put failed` errors were observed in the cleaned run.

Performance comparison on the same long-form request:

| Version | completed/failed | audio_s | mean e2e ms | median e2e ms | mean RTF | median RTF | throughput | peak VRAM MiB|
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| `ff6f0da2` baseline | 3/0 | 80.0 | 52518.83 | 48050.04 | 0.6565 | 0.6006 | 1.5443 | 72027 |
| final cleaned commit | 3/0 | 80.0 | 49772.06 | 45565.51 | 0.6222 | 0.5696 | 1.6311 | 72027 |

Final improvement:

- vs `ff6f0da2` baseline:
  - mean e2e latency improved by `5.23%`
  - median e2e latency improved by `5.17%`
  - audio throughput improved by `5.62%`

Audio-quality comparison was run on paired baseline-vs-PR outputs generated from the same long-form request and reference audio. Both outputs are 80.0s stereo PCM.

| Metric | Result |
|---|---:|
| MCD over MFCC[1:13] with DTW, 16 kHz mono | `0.0291 dB` |
| PCM correlation | `0.9999999993` |
| PCM SNR vs baseline | `88.60 dB` |
| Max absolute PCM sample delta | `6 / 32767` |
| Mean absolute PCM sample delta | `0.0122` |
| RMS sample delta | `0.1125` |

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/distributed/omni_connectors/utils/serialization.py [hunk 1]
```diff
@@ -87,9 +87,7 @@ def _encode_tensor(self, tensor: torch.Tensor) -> dict[str, Any]:
         # Handle 0-dimensional (scalar) tensors by reshaping to 1D first
         if t.dim() == 0:
             t = t.reshape(1)
-        if not t.is_contiguous():
-            t = t.contiguous()
-        t = t.view(torch.uint8)
+        t = t.reshape(-1).contiguous().view(torch.uint8)
```
Comment (lishunyang12) [inline vllm_omni/distributed/omni_connectors/utils/serialization.py]: `reshape(-1)` already gives back a contiguous 1D tensor, so the `.contiguous()` here is always a no-op. Just do `t.reshape(-1).view(torch.uint8)`.

File: vllm_omni/model_executor/models/moss_tts/modeling_moss_tts_codec.py [hunk 2]
```diff
@@ -34,6 +34,61 @@
...
+    source = payload if payload is not None else input_seg
+    if isinstance(source, torch.Tensor):
+        codes_tensor = source
+    elif isinstance(source, (list, tuple)):
+        codes_tensor = torch.tensor(source, dtype=torch.long)
+    else:
+        return None
+
+    if codes_tensor.ndim == 2:
+        if int(codes_tensor.shape[0]) == int(n_vq):
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/moss_tts/modeling_moss_tts_codec.py]: When a chunk carries exactly `n_vq` frames the payload is `[n_vq, n_vq]` and this always treats shape[0] as the codebook dim, so a frame-major producer of that size gets read transposed. Key off the source path instead of inferring layout from shape.

File: vllm_omni/model_executor/models/moss_tts/modeling_moss_tts_talker.py [hunk 3]
```diff
@@ -74,6 +74,135 @@ def _iter_state_row_spans(
...
+    return [[row[cb] for row in tail] for cb in range(n_vq)]
+
+
+def _moss_local_append_history_frame(
+    audio_codes: dict[str, Any],
+    new_codes: torch.Tensor,
+    n_vq: int,
+) -> dict[str, Any]:
+    accumulated = audio_codes.get("accumulated")
+    rep_window = 50
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/moss_tts/modeling_moss_tts_talker.py]: `rep_window = 50` is hardcoded here and again in `write_update`. Pull it into a module-level `_MOSS_LOCAL_REP_WINDOW = 50` so the two can't drift.

**Review by lishunyang12 — COMMENTED** [review]
Minor mostly — the `n_vq x n_vq` transpose edge case is the only one to actually check.

File: vllm_omni/model_executor/stage_input_processors/moss_tts.py [hunk 4]
```diff
@@ -258,13 +313,24 @@ def talker2codec_raw_async_chunk(
...
             for frame in frames_cpu[valid_rows]:
-                pending_frames.append(frame.clone())
+                new_frames_cpu.append(frame.clone().contiguous())
         # Raw/local streaming should mirror the non-streaming path: the codec
         # decodes only generated audio rows. Reference audio conditions the
         # talker, but feeding its codes into the codec streaming state adds a
         # long first-packet prime step and changes the decoder state relative
         # to non-streaming output.
 
+    state = _get_moss_raw_chunk_state(transfer_manager, req_id)
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/stage_input_processors/moss_tts.py:323]: This per-request state has no abort/error cleanup: `cleanup_sender` clears `code_prompt_token_ids` and `request_payload`, but not `_moss_tts_raw_chunk_states`, and pruning only removes finalized entries. A request cancelled before a final callback leaves its pending tensors indefinitely (and reusing the external ID can mix stale frames). Please clear this state from the transfer-manager cleanup path.

File: vllm_omni/model_executor/stage_input_processors/moss_tts.py [hunk 5]
```diff
@@ -283,58 +349,56 @@ def talker2codec_raw_async_chunk(
...
+    state.emitted_frame_count += emit_frames
+    finished = bool(is_finished and len(state.pending_frames) == 0)
 
     if finished:
+        state.final_flushed = True
         transfer_manager.code_prompt_token_ids.pop(req_id, None)
         transfer_manager.request_payload.pop(req_id, None)
+        _prune_moss_raw_finalized_states(transfer_manager)
+
+    codec_codes = chunk_codes.transpose(0, 1).contiguous().to(torch.long)
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/stage_input_processors/moss_tts.py:391]: Please add baseline-vs-PR audio-quality evidence for this `[Perf]` change (same inputs/seeds, with a metric such as MCD/WER/embedding similarity or paired audio samples). The description reports completion/latency/throughput, but does not rule out a quality regression from the new codec transport/layout. It also lacks warmup exclusion, mean±stddev, and peak-VRAM measurements for the performance comparison.

File: vllm_omni/model_executor/models/moss_tts/modeling_moss_tts_codec.py (hunk 2 above)
Comment (nagisa-kunhah) [reply vllm_omni/model_executor/models/moss_tts/modeling_moss_tts_codec.py]: Thanks! Fixed by making the payload shape explicit as codec-native `[n_vq, T]`.

File: vllm_omni/model_executor/stage_input_processors/moss_tts.py (hunk 4 above)
Comment (nagisa-kunhah) [reply vllm_omni/model_executor/stage_input_processors/moss_tts.py:323]: Thanks, added cleanup for `_moss_tts_raw_chunk_states` in both `OmniChunkTransferAdapter` and `OmniConnectorModelRunnerMixin`.

File: vllm_omni/model_executor/stage_input_processors/moss_tts.py (hunk 5 above)
Comment (nagisa-kunhah) [reply vllm_omni/model_executor/stage_input_processors/moss_tts.py:391]: Thanks! I add the baseline-vs-PR audio-quality evidence to the PR description. and the metrics show no observable audio-quality regression from the codec transport/layout change.
